ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/LagReportingAgent.java

```diff
@@ -0,0 +1,391 @@
+      receivedLagInfo;
+  private final Map<HostInfo, Boolean> aliveHosts;
+
+  private URL localURL;
+
+  /**
+   * Builder for creating an instance of LagReportingAgent.
+   * @return
+   */
+  public static LagReportingAgent.Builder builder() {
+    return new LagReportingAgent.Builder();
+  }
+
+  /**
+   * Lag related agent for both sending out lag for localhost as well as receiving lag reports for
+   * other nodes in the cluster.
+   * @param engine The ksql engine to access streams for inferring the cluster and getting local
+   *               lags metrics
+   * @param serviceContext Service context for issuing ksql requests
+   * @param config The LagReportingConfig for configuring this agent
+   * @param clock Clock for reporting lag
+   */
+  private LagReportingAgent(
+      final KsqlEngine engine,
+      final ScheduledExecutorService scheduledExecutorService,
+      final ServiceContext serviceContext,
+      final LagReportingConfig config,
+      final Clock clock) {
+    this.engine = requireNonNull(engine, "engine");
+    this.scheduledExecutorService = scheduledExecutorService;
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.config = requireNonNull(config, "configuration parameters");
+    this.clock = clock;
+    this.serviceManager = new ServiceManager(Arrays.asList(new SendLagService()));
+    this.receivedLagInfo = new ConcurrentHashMap<>();
+    this.aliveHosts = new ConcurrentHashMap<>();
+  }
+
+  void setLocalAddress(final String applicationServer) {
+    try {
+      this.localURL = new URL(applicationServer);
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to convert remote host info to URL."
+          + " remoteInfo: " + applicationServer);
+    }
+  }
+
+  void startAgent() {
+    try {
+      serviceManager.startAsync().awaitHealthy(SERVICE_TIMEOUT_SEC, TimeUnit.SECONDS);
+    } catch (TimeoutException | IllegalStateException e) {
+      LOG.error("Failed to start heartbeat services with exception " + e.getMessage(), e);
+    }
+  }
+
+  void stopAgent() {
+    try {
+      serviceManager.stopAsync().awaitStopped(SERVICE_TIMEOUT_SEC, TimeUnit.SECONDS);
+    } catch (TimeoutException | IllegalStateException e) {
+      LOG.error("Failed to stop heartbeat services with exception " + e.getMessage(), e);
+    } finally {
+      scheduledExecutorService.shutdownNow();
+    }
+  }
+
+  /**
+   * Stores the host lag received from a remote Ksql server.
+   * @param lagReportingRequest The host lag information sent directly from the other node.
+   */
+  public void receiveHostLag(final LagReportingRequest lagReportingRequest) {
+    final long nowMs = clock.millis();
+    final HostInfoEntity hostInfoEntity = lagReportingRequest.getHostInfo();
+    final HostInfo hostInfo = new HostInfo(hostInfoEntity.getHost(), hostInfoEntity.getPort());
+    receivedLagInfo.remove(hostInfo);
```

Comment:
Why do you remove the entry here? All you want to do is update the offsets and lag information if the current request has newer information. Keep in mind, requests may arrive out of order

Reply:
I remove the entry so as to produce one consistent view of the data, per host.  I could check the timestamps and ignore if it's an older request, if that's what you mean.  I'll do that.

If you don't remove data for this host, then over time, it will be littered with state store that no longer exist or partitions that the host no longer supports.  Maybe this isn't a huge deal, but why not be consistent.